### PR TITLE
Solve issue with static function and update versions of packages in CI

### PR DESCRIPTION
**Pull request by ArseniyKholod**
As suggested in https://github.com/JuliaInterop/libcxxwrap-julia/issues/150, I correct the way to wrap static functions, it is able now to be compiled with `libcxxwrap-julia` v0.12.2

### CODE REVIEW

**Review by sloede — COMMENTED** [review]
LGTM, thanks a lot for fixing this! One small comment, otherwise good to merge (once the tests pass)

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -26,7 +26,7 @@ env:
   # Modify these variables for upstream package versions - do NOT hardcode the version
   # anywhere else!
   OPENFHE_VERSION: 1.1.2
```
[the remark this answers is not in the document]
Comment (ArseniyKholod) [reply .github/workflows/ci.yml]: Is it possible to rewrite compat settings for `OpenFHE.jl` localy on CI for downstream test? For example for `CxxWrap.jl` v0.15.

[the remark this answers is not in the document]
Comment (ArseniyKholod) [reply .github/workflows/ci.yml]: Anyway, it's necessary to solve issue with static functions, so PR could be merged, but I'm not sure what to do with `OpenFHE` and `libcxxwar-julia` versions in CI. What would you say?

[the remark this answers is not in the document]
Comment (sloede) [reply .github/workflows/ci.yml]: Can you please try to apply the following patch and see if that fixes the issue?
```diff
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -25,9 +25,10 @@ on:
 env:
   # Modify these variables for upstream package versions - do NOT hardcode the version
   # anywhere else!
-  OPENFHE_VERSION: 1.1.4
-  LIBCXXWRAP_JULIA_VERSION: 0.12.2
-  JULIA_VERSION: 1.9
+  OPENFHE_VERSION: '1.1.4'
+  LIBCXXWRAP_JULIA_VERSION: '0.12.2'
+  JULIA_VERSION: '1.9'
+  FORCE_CXXWRAP_JL_VERSION: '0.15' # Use only briefly during transition (default: '')

 jobs:
   test:
@@ -151,6 +152,9 @@ jobs:
         run: |
           cp test/LocalPreferences.toml test/OpenFHE.jl/LocalPreferences.toml
           sed -i "s/MyOpenFHE/OpenFHE/" test/OpenFHE.jl/LocalPreferences.toml
+          if [ -n "$FORCE_CXXWRAP_JL_VERSION" ]; then
+            sed -i "s/CxxWrap = \"[^\"]*\"/CxxWrap = \"=$FORCE_CXXWRAP_JL_VERSION\"/" test/OpenFHE.jl/Project.toml
+          fi
       - name: Run downstream Julia tests
         if: ${{ matrix.os == 'ubuntu-latest' }}
         shell: julia --color=yes --project=test/OpenFHE.jl {0}
```
If it does, we can merge it and then just have to remember to reset the `FORCE_CXXWRAP_JL_VERSION` variable to an empty string once OpenFHE.jl has been updated.

**Comment by sloede** [comment]
I'm not sure I see the problem here. It *looks* like we force the version of CxxWrap.jl to v0.15 too early, but I don't know why - the `LocalPreferences.toml` file should already take care of this, I believe.

Maybe you need - for debugging - enable ssh debugging unconditionally by changing
https://github.com/sloede/openfhe-julia/blob/aeeb05b1300ac9f334d1fb6a463316a01e11f9f9/.github/workflows/ci.yml#L172
to something like
```yml
if: {{ always() }}
```
Then wait until the job has reached that stage, then go to the logs, copy the ssh link, and log onto the GitHub runner.

**Comment by ArseniyKholod** [comment]
> I'm not sure I see the problem here. It _looks_ like we force the version of CxxWrap.jl to v0.15 too early, but I don't know why - the `LocalPreferences.toml` file should already take care of this, I believe.
> 
> Maybe you need - for debugging - enable ssh debugging unconditionally by changing
> 
> https://github.com/sloede/openfhe-julia/blob/aeeb05b1300ac9f334d1fb6a463316a01e11f9f9/.github/workflows/ci.yml#L172
> 
> 
> to something like
> ```yaml
> if: {{ always() }}
> ```
> 
> Then wait until the job has reached that stage, then go to the logs, copy the ssh link, and log onto the GitHub runner.

Thanks!!
Locally on runner I was able to reproduce new error mentioned [here](https://github.com/sloede/OpenFHE.jl/issues/43#issuecomment-2059099346).
In downstream check happens smth really strange, because it utilizes not the compiled `openfhe-julia` from this PR, but `openfhe_julia_jll` 
https://github.com/sloede/openfhe-julia/actions/runs/8706416826/job/23878936610?pr=35

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -151,6 +152,11 @@ jobs:
         run: |
           cp test/LocalPreferences.toml test/OpenFHE.jl/LocalPreferences.toml
           sed -i "s/MyOpenFHE/OpenFHE/" test/OpenFHE.jl/LocalPreferences.toml
+          if [ -n "$FORCE_CXXWRAP_JL_VERSION" ]; then
+            sed -i "s/CxxWrap = \"0[^\"]*\"/CxxWrap = \"=$FORCE_CXXWRAP_JL_VERSION\"/" test/OpenFHE.jl/Project.toml
+            sed -i "s/using openfhe_julia_jll: libopenfhe_julia/#using openfhe_julia_jll: libopenfhe_julia/" test/OpenFHE.jl/src/OpenFHE.jl
+            sed -i "s/const libopenfhe_julia_path = @load_preference(\"libopenfhe_julia\", libopenfhe_julia)/const libopenfhe_julia_path = @load_preference(\"libopenfhe_julia\")#, libopenfhe_julia)/" test/OpenFHE.jl/src/OpenFHE.jl
```
[the remark this answers is not in the document]
Comment (ArseniyKholod) [reply .github/workflows/ci.yml:158]: It works, but I don't know actually why `openfhe_julia_jll` creates problem, when using custom openfhe library, sounds not logical for me...

[the remark this answers is not in the document]
Comment (sloede) [reply .github/workflows/ci.yml:158]: Hm, I'm hesitant to merge while we do not understand the underlying issue. Have you tried reproducing the steps on roci?

In general, yes, we can go ahead and merge this, assuming that all will be well in due time once all packages are updated to their latest version. It does bother me, though, that we do not know _why_ this isn't working (when we think it should). Probably the error is on our (my) side, but I'd like to be sure of that 😅

[the remark this answers is not in the document]
Comment (sloede) [reply .github/workflows/ci.yml:158]: I think (I *think*) that I'm starting to wrap my head around this:
* We build local, updated versions of OpenFHE, libcxxwrap-julia, openfhe-julia
* In the downstream tests, we install OpenFHE.jl, with the custom, locally built `libopenfhe_julia.so` is selected for loading using Preferences.jl
* *However*, before the preferences are loaded, the `openfhe_julia_jll`-provided `libopenfhe_julia.so` library from the JLL library is loaded, *which in turn loads its own `libcxxwrap_julia.so` library* (plus OpenFHE etc.).

To verify this, I performed the following steps from a project with the installation of the regular JLL package:
```julia
julia> using openfhe_julia_jll

julia> using Libdl

julia> dlopen("/home/mschlott/hackathon/fhe/mwe/ak-openfhe-julia/build/lib/libopenfhe_julia.so")
ERROR: could not load library "/home/mschlott/hackathon/fhe/mwe/ak-openfhe-julia/build/lib/libopenfhe_julia.so"
/home/mschlott/hackathon/fhe/mwe/ak-openfhe-julia/build/lib/libopenfhe_julia.so: undefined symbol: _ZN8lbcrypto12DCRTPolyImplIN9bigintdyn9mubintvecINS1_5ubintImEEEEE23DropLastElementAndScaleERKSt6vectorIN6intnat14NativeIntegerTImEESaISA_EESE_
Stacktrace:
 [1] dlopen(s::String, flags::UInt32; throw_error::Bool)
   @ Base.Libc.Libdl ./libdl.jl:117
 [2] dlopen
   @ Base.Libc.Libdl ./libdl.jl:116 [inlined]
 [3] dlopen(s::String)
   @ Base.Libc.Libdl ./libdl.jl:116
 [4] top-level scope
   @ REPL[3]:1
```
which yielded the same error you reported. Thus, my conclusion here is that it is too much of a hassle to fix this upstream test for the (comparatively) rare case of multiple incompatible libraries, _especially_ given that the update in this PR is about to fix them 🙂

Comment (sloede) [inline .github/workflows/ci.yml:158]:
```suggestion
            # TODO: the following two lines are a workaround for using an incompatible JLL package
            sed -i "s/using openfhe_julia_jll: libopenfhe_julia/#using openfhe_julia_jll: libopenfhe_julia/" test/OpenFHE.jl/src/OpenFHE.jl
            sed -i "s/const libopenfhe_julia_path = @load_preference(\"libopenfhe_julia\", libopenfhe_julia)/const libopenfhe_julia_path = @load_preference(\"libopenfhe_julia\")#, libopenfhe_julia)/" test/OpenFHE.jl/src/OpenFHE.jl
```

**Review by sloede — APPROVED** [review]
LGTM!

**Pull request merged**